README.rst
WARNING: This is a almost a copy of scrapy-slash but works with https://github.com/prerender/prerender

==============================================
Scrapy & JavaScript integration through Prerender
==============================================
.. _Scrapy: https://github.com/scrapy/scrapy

Installation
============

Install scrapy-prerender::

    $ python setup.py install


Configuration
=============

1. Add the Prerender server address to ``settings.py`` of your Scrapy project
   like this::

      PRERENDER_URL = 'http://192.168.59.103:8050'

2. Enable the Prerender middleware by adding it to ``DOWNLOADER_MIDDLEWARES``
   in your ``settings.py`` file and changing HttpCompressionMiddleware
   priority::

      DOWNLOADER_MIDDLEWARES = {
          'scrapy_prerender.PrerenderCookiesMiddleware': 723,
          'scrapy_prerender.PrerenderMiddleware': 725,
          'scrapy.downloadermiddlewares.httpcompression.HttpCompressionMiddleware': 810,
      }

   Order `723` is just before `HttpProxyMiddleware` (750) in default
   scrapy settings.

   HttpCompressionMiddleware priority should be changed in order to allow
   advanced response processing; see https://github.com/scrapy/scrapy/issues/1895
   for details.

3. Enable ``PrerenderDeduplicateArgsMiddleware`` by adding it to
   ``SPIDER_MIDDLEWARES`` in your ``settings.py``::

      SPIDER_MIDDLEWARES = {
          'scrapy_prerender.PrerenderDeduplicateArgsMiddleware': 100,
      }

   This middleware is needed to support ``cache_args`` feature; it allows
   to save disk space by not storing duplicate Prerender arguments multiple
   times in a disk request queue. If Prerender 2.1+ is used the middleware
   also allows to save network traffic by not sending these duplicate
   arguments to Prerender server multiple times.

4. Set a custom ``DUPEFILTER_CLASS``::

      DUPEFILTER_CLASS = 'scrapy_prerender.PrerenderAwareDupeFilter'

5. If you use Scrapy HTTP cache then a custom cache storage backend
   is required. scrapy-prerender provides a subclass of
   ``scrapy.contrib.httpcache.FilesystemCacheStorage``::

      HTTPCACHE_STORAGE = 'scrapy_prerender.PrerenderAwareFSCacheStorage'

   If you use other cache storage then it is necesary to subclass it and
   replace all ``scrapy.util.request.request_fingerprint`` calls with
   ``scrapy_prerender.prerender_request_fingerprint``.

.. note::

    Steps (4) and (5) are necessary because Scrapy doesn't provide a way
    to override request fingerprints calculation algorithm globally; this
    could change in future.


There are also some additional options available.
Put them into your ``settings.py`` if you want to change the defaults:

* ``PRERENDER_COOKIES_DEBUG`` is ``False`` by default.
  Set to ``True`` to enable debugging cookies in the ``PrerenderCookiesMiddleware``.
  This option is similar to ``COOKIES_DEBUG``
  for the built-in scarpy cookies middleware: it logs sent and received cookies
  for all requests.
* ``PRERENDER_LOG_400`` is ``True`` by default - it instructs to log all 400 errors
  from Prerender. They are important because they show errors occurred
  when executing the Prerender script. Set it to ``False`` to disable this logging.
* ``PRERENDER_SLOT_POLICY`` is ``scrapy_prerender.SlotPolicy.PER_DOMAIN`` (as object, not just a string) by default.
  It specifies how concurrency & politeness are maintained for Prerender requests,
  and specify the default value for ``slot_policy`` argument for
  ``PrerenderRequest``, which is described below.


Usage
=====

Requests
--------

The easiest way to render requests with Prerender is to
use ``scrapy_prerender.PrerenderRequest``::

    yield PrerenderRequest(url, self.parse_result,
        args={
            # optional; parameters passed to Prerender HTTP API
            'wait': 0.5,

            # 'url' is prefilled from request url
            # 'http_method' is set to 'POST' for POST requests
            # 'body' is set to request body for POST requests
        },
        endpoint='render.json', # optional; default is render
        prerender_url='<url>',     # optional; overrides PRERENDER_URL
        slot_policy=scrapy_prerender.SlotPolicy.PER_DOMAIN,  # optional
    )

Alternatively, you can use regular scrapy.Request and
``'prerender'`` Request `meta` key::

    yield scrapy.Request(url, self.parse_result, meta={
        'prerender': {
            'args': {
                # set rendering arguments here
                'html': 1,
                'png': 1,

                # 'url' is prefilled from request url
                # 'http_method' is set to 'POST' for POST requests
                # 'body' is set to request body for POST requests
            },

            # optional parameters
            'endpoint': 'render.json',  # optional; default is render.json
            'prerender_url': '<url>',      # optional; overrides PRERENDER_URL
            'slot_policy': scrapy_prerender.SlotPolicy.PER_DOMAIN,
            'prerender_headers': {},       # optional; a dict with headers sent to Prerender
            'dont_process_response': True, # optional, default is False
            'dont_send_headers': True,  # optional, default is False
            'magic_response': False,    # optional, default is True
        }
    })

Use ``request.meta['prerender']`` API in middlewares or when scrapy.Request
subclasses are used (there is also ``PrerenderFormRequest`` described below).
For example, ``meta['prerender']`` allows to create a middleware which enables
Prerender for all outgoing requests by default.

``PrerenderRequest`` is a convenient utility to fill ``request.meta['prerender']``;
it should be easier to use in most cases. For each ``request.meta['prerender']``
key there is a corresponding ``PrerenderRequest`` keyword argument: for example,
to set ``meta['prerender']['args']`` use ``PrerenderRequest(..., args=myargs)``.

* ``meta['prerender']['args']`` contains arguments sent to Prerender.
  scrapy-prerender adds some default keys/values to ``args``:

  * 'url' is set to request.url;
  * 'http_method' is set to 'POST' for POST requests;
  * 'body' is set to to request.body for POST requests.

  You can override default values by setting them explicitly.

  Note that by default Scrapy escapes URL fragments using AJAX escaping scheme.
  If you want to pass a URL with a fragment to Prerender then set ``url``
  in ``args`` dict manually. This is handled automatically if you use
  ``PrerenderRequest``, but you need to keep that in mind if you use raw
  ``meta['prerender']`` API.

  Prerender 1.8+ is required to handle POST requests; in earlier Prerender versions
  'http_method' and 'body' arguments are ignored. If you work with ``/execute``
  endpoint and want to support POST requests you have to handle
  ``http_method`` and ``body`` arguments in your Lua script manually.

* ``meta['prerender']['cache_args']`` is a list of argument names to cache
  on Prerender side. These arguments are sent to Prerender only once, then cached
  values are used; it allows to save network traffic and decreases request
  queue disk memory usage. Use ``cache_args`` only for large arguments
  which don't change with each request; ``lua_source`` is a good candidate
  (if you don't use string formatting to build it). Prerender 2.1+ is required
  for this feature to work.

* ``meta['prerender']['endpoint']`` is the Prerender endpoint to use.
  In case of PrerenderRequest
  `render <http://prerender.readthedocs.org/en/latest/api.html#render-html>`_
  is used by default. If you're using raw scrapy.Request then
  `render.json <http://prerender.readthedocs.org/en/latest/api.html#render-json>`_
  is a default (for historical reasons). It is better to always pass endpoint
  explicitly.

  See Prerender `HTTP API docs`_ for a full list of available endpoints
  and parameters.

.. _HTTP API docs: http://prerender.readthedocs.org/en/latest/api.html

* ``meta['prerender']['prerender_url']`` overrides the Prerender URL set
  in ``settings.py``.

* ``meta['prerender']['prerender_headers']`` allows to add or change headers
  which are sent to Prerender server. Note that this option **is not** for
  setting headers which are sent to the remote website.

* ``meta['prerender']['slot_policy']`` customize how
  concurrency & politeness are maintained for Prerender requests.

  Currently there are 3 policies available:

  1. ``scrapy_prerender.SlotPolicy.PER_DOMAIN`` (default) - send Prerender requests to
     downloader slots based on URL being rendered. It is useful if you want
     to maintain per-domain politeness & concurrency settings.

  2. ``scrapy_prerender.SlotPolicy.SINGLE_SLOT`` - send all Prerender requests to
     a single downloader slot. It is useful if you want to throttle requests
     to Prerender.

  3. ``scrapy_prerender.SlotPolicy.SCRAPY_DEFAULT`` - don't do anything with slots.
     It is similar to ``SINGLE_SLOT`` policy, but can be different if you access
     other services on the same address as Prerender.

* ``meta['prerender']['dont_process_response']`` - when set to True,
  PrerenderMiddleware won't change the response to a custom scrapy.Response
  subclass. By default for Prerender requests one of PrerenderResponse,
  PrerenderTextResponse or PrerenderJsonResponse is passed to the callback.

* ``meta['prerender']['dont_send_headers']``: by default scrapy-prerender passes
  request headers to Prerender in 'headers' JSON POST field. For all render.xxx
  endpoints it means Scrapy header options are respected by default
  (http://prerender.readthedocs.org/en/stable/api.html#arg-headers). In Lua
  scripts you can use ``headers`` argument of ``prerender:go`` to apply the
  passed headers: ``prerender:go{url, headers=prerender.args.headers}``.

  Set 'dont_send_headers' to True if you don't want to pass ``headers``
  to Prerender.

* ``meta['prerender']['http_status_from_error_code']`` - set response.status
  to HTTP error code when ``assert(prerender:go(..))`` fails; it requires
  ``meta['prerender']['magic_response']=True``. ``http_status_from_error_code``
  option is False by default if you use raw meta API;
  PrerenderRequest sets it to True by default.

* ``meta['prerender']['magic_response']`` - when set to True and a JSON
  response is received from Prerender, several attributes of the response
  (headers, body, url, status code) are filled using data returned in JSON:

  * response.headers are filled from 'headers' keys;
  * response.url is set to the value of 'url' key;
  * response.body is set to the value of 'html' key,
    or to base64-decoded value of 'body' key;
  * response.status is set to the value of 'http_status' key.
    When ``meta['prerender']['http_status_from_error_code']`` is True
    and ``assert(prerender:go(..))`` fails with an HTTP error
    response.status is also set to HTTP error code.

  Original URL, status and headers are available as ``response.real_url``,
  ``response.prerender_response_status`` and ``response.prerender_response_headers``.

  This option is set to True by default if you use PrerenderRequest.
  ``render.json`` and ``execute`` endpoints may not have all the necessary
  keys/values in the response.
  For non-JSON endpoints, only url is filled, regardless of the
  ``magic_response`` setting.


Use ``scrapy_prerender.PrerenderFormRequest`` if you want to make a ``FormRequest``
via prerender. It accepts the same arguments as ``PrerenderRequest``,
and also ``formdata``, like ``FormRequest`` from scrapy::

    >>> PrerenderFormRequest('http://example.com', formdata={'foo': 'bar'})
    <POST http://example.com>

``PrerenderFormRequest.from_response`` is also supported, and works as described
in `scrapy documentation <http://scrapy.readthedocs.org/en/latest/topics/request-response.html#scrapy.http.FormRequest.from_response>`_.

Responses
---------

scrapy-prerender returns Response subclasses for Prerender requests:

* PrerenderResponse is returned for binary Prerender responses - e.g. for
  /render.png responses;
* PrerenderTextResponse is returned when the result is text - e.g. for
  /render responses;
* PrerenderJsonResponse is returned when the result is a JSON object - e.g.
  for /render.json responses or /execute responses when script returns
  a Lua table.

To use standard Response classes set ``meta['prerender']['dont_process_response']=True``
or pass ``dont_process_response=True`` argument to PrerenderRequest.

All these responses set ``response.url`` to the URL of the original request
(i.e. to the URL of a website you want to render), not to the URL of the
requested Prerender endpoint. "True" URL is still available as
``response.real_url``.

PrerenderJsonResponse provide extra features:

* ``response.data`` attribute contains response data decoded from JSON;
  you can access it like ``response.data['html']``.

* If Prerender session handling is configured, you can access current cookies
  as ``response.cookiejar``; it is a CookieJar instance.

* If Scrapy-Prerender response magic is enabled in request (default),
  several response attributes (headers, body, url, status code)
  are set automatically from original response body:

  * response.headers are filled from 'headers' keys;
  * response.url is set to the value of 'url' key;
  * response.body is set to the value of 'html' key,
    or to base64-decoded value of 'body' key;
  * response.status is set from the value of 'http_status' key.


CHANGES.rst
Changes
=======

0.7.2 (2017-03-30)
------------------

* fixed issue with response type detection.

0.7.1 (2016-12-20)
------------------

* Scrapy 1.0.x support is back;
* README updates.

0.7 (2016-05-16)
----------------

* ``PRERENDER_COOKIES_DEBUG`` setting allows to log cookies
  sent and received to/from Prerender in ``cookies`` request/response fields.
  It is similar to Scrapy's builtin ``COOKIES_DEBUG``, but works for
  Prerender requests;
* README cleanup.

0.6.1 (2016-04-29)
------------------

* Warning about HTTP methods is no longer logged for non-Prerender requests.

0.6 (2016-04-20)
----------------

* ``PrerenderAwareDupeFilter`` and ``prerender_request_fingerprint`` are improved:
  they now canonicalize URLs and take URL fragments in account;
* ``cache_args`` value fingerprints are now calculated faster.

0.5 (2016-04-18)
----------------

* ``cache_args`` PrerenderRequest argument and
  ``request.meta['prerender']['cache_args']`` key allow to save network traffic
  and disk storage by not storing duplicate Prerender arguments in disk request
  queues and not sending them to Prerender multiple times. This feature requires
  Prerender 2.1+.

To upgrade from v0.4 enable ``PrerenderDeduplicateArgsMiddleware`` in settings.py::

  SPIDER_MIDDLEWARES = {
      'scrapy_prerender.PrerenderDeduplicateArgsMiddleware': 100,
  }

0.4 (2016-04-14)
----------------

* PrerenderFormRequest class is added; it is a variant of FormRequest which uses
  Prerender;
* Prerender parameters are no longer stored in request.meta twice; this change
  should decrease disk queues data size;
* PrerenderMiddleware now increases request priority when rescheduling the request;
  this should decrease disk queue data size and help with stale cookie
  problems.

0.3 (2016-04-11)
----------------

Package is renamed from ``scrapyjs`` to ``scrapy-prerender``.

An easiest way to upgrade is to replace ``scrapyjs`` imports with
``scrapy_prerender`` and update ``settings.py`` with new defaults
(check the README).

There are many new helpers to handle JavaScript rendering transparently;
the recommended way is now to use ``scrapy_prerender.PrerenderRequest`` instead
of  ``request.meta['prerender']``. Please make sure to read the README if
you're upgrading from scrapyjs - you may be able to drop some code from your
project, especially if you want to access response html, handle cookies
and headers.

* new PrerenderRequest class; it can be used as a replacement for scrapy.Request
  to provide a better integration with Prerender;
* added support for POST requests;
* PrerenderResponse, PrerenderTextResponse and PrerenderJsonResponse allow to
  handle Prerender responses transparently, taking care of response.url,
  response.body, response.headers and response.status. PrerenderJsonResponse
  allows to access decoded response JSON data as ``response.data``.
* cookie handling improvements: it is possible to handle Scrapy and Prerender
  cookies transparently; current cookiejar is exposed as response.cookiejar;
* headers are passed to Prerender by default;
* URLs with fragments are handled automatically when using PrerenderRequest;
* logging is improved: ``PrerenderRequest.__repr__`` shows both requested URL
  and Prerender URL;
* in case of Prerender HTTP 400 errors the response is logged by default;
* an issue with dupefilters is fixed: previously the order of keys in
  JSON request body could vary, making requests appear as non-duplicates;
* it is now possible to pass custom headers to Prerender server itself;
* test coverage reports are enabled.

0.2 (2016-03-26)
----------------

* Scrapy 1.0 and 1.1 support;
* Python 3 support;
* documentation improvements;
* project is moved to https://github.com/scrapy-plugins/scrapy-prerender.

0.1.1 (2015-03-16)
------------------

Fixed fingerprint calculation for non-string meta values.

0.1 (2015-02-28)
----------------

Initial release


scrapy_prerender/middleware.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import

import copy
import json
import logging
import warnings
from collections import defaultdict

from six.moves.urllib.parse import urljoin
from six.moves.http_cookiejar import CookieJar

import scrapy
from scrapy.exceptions import NotConfigured
from scrapy.http.headers import Headers
from scrapy.http.response.text import TextResponse
from scrapy import signals

from scrapy_prerender.responsetypes import responsetypes
from scrapy_prerender.cookies import jar_to_har, har_to_jar
from scrapy_prerender.utils import (
    scrapy_headers_to_unicode_dict,
    json_based_hash,
    parse_x_prerender_saved_arguments_header,
)
from scrapy_prerender.response import get_prerender_status, get_prerender_headers


logger = logging.getLogger(__name__)


class SlotPolicy(object):
    PER_DOMAIN = 'per_domain'
    SINGLE_SLOT = 'single_slot'
    SCRAPY_DEFAULT = 'scrapy_default'

    _known = {PER_DOMAIN, SINGLE_SLOT, SCRAPY_DEFAULT}


class PrerenderCookiesMiddleware(object):
    """
    This downloader middleware maintains cookiejars for Prerender requests.

    It gets cookies from 'cookies' field in Prerender JSON responses
    and sends current cookies in 'cookies' JSON POST argument instead of
    sending them in http headers.

    It should process requests before PrerenderMiddleware, and process responses
    after PrerenderMiddleware.
    """
    def __init__(self, debug=False):
        self.jars = defaultdict(CookieJar)
        self.debug = debug

    @classmethod
    def from_crawler(cls, crawler):
        return cls(debug=crawler.settings.getbool('PRERENDER_COOKIES_DEBUG'))

    def process_request(self, request, spider):
        """
        For Prerender requests add 'cookies' key with current
        cookies to ``request.meta['prerender']['args']`` and remove cookie
        headers sent to Prerender itself.
        """
        if 'prerender' not in request.meta:
            return

        if request.meta.get('_prerender_processed'):
            request.headers.pop('Cookie', None)
            return

        prerender_options = request.meta['prerender']

        prerender_args = prerender_options.setdefault('args', {})
        if 'cookies' in prerender_args:  # cookies already set
            return

        if 'session_id' not in prerender_options:
            return

        jar = self.jars[prerender_options['session_id']]

        cookies = self._get_request_cookies(request)
        har_to_jar(jar, cookies)

        prerender_args['cookies'] = jar_to_har(jar)
        self._debug_cookie(request, spider)

    def process_response(self, request, response, spider):
        """
        For Prerender JSON responses add all cookies from
        'cookies' in a response to the cookiejar.
        """
        from scrapy_prerender import PrerenderJsonResponse
        if not isinstance(response, PrerenderJsonResponse):
            return response

        if 'cookies' not in response.data:
            return response

        if 'prerender' not in request.meta:
            return response

        if not request.meta.get('_prerender_processed'):
            warnings.warn("PrerenderCookiesMiddleware requires PrerenderMiddleware")
            return response

        prerender_options = request.meta['prerender']
        session_id = prerender_options.get('new_session_id',
                                        prerender_options.get('session_id'))
        if session_id is None:
            return response

        jar = self.jars[session_id]
        request_cookies = prerender_options['args'].get('cookies', [])
        har_to_jar(jar, response.data['cookies'], request_cookies)
        self._debug_set_cookie(response, spider)
        response.cookiejar = jar
        return response

    def _get_request_cookies(self, request):
        if isinstance(request.cookies, dict):
            return [
                {'name': k, 'value': v} for k, v in request.cookies.items()
            ]
        return request.cookies or []

    def _debug_cookie(self, request, spider):
        if self.debug:
            cl = request.meta['prerender']['args']['cookies']
            if cl:
                cookies = '\n'.join(
                    'Cookie: {}'.format(self._har_repr(c)) for c in cl)
                msg = 'Sending cookies to: {}\n{}'.format(request, cookies)
                logger.debug(msg, extra={'spider': spider})

    def _debug_set_cookie(self, response, spider):
        if self.debug:
            cl = response.data['cookies']
            if cl:
                cookies = '\n'.join(
                    'Set-Cookie: {}'.format(self._har_repr(c)) for c in cl)
                msg = 'Received cookies from: {}\n{}'.format(response, cookies)
                logger.debug(msg, extra={'spider': spider})

    @staticmethod
    def _har_repr(har_cookie):
        return '{}={}'.format(har_cookie['name'], har_cookie['value'])


class PrerenderDeduplicateArgsMiddleware(object):
    """
    Spider middleware which allows not to store duplicate Prerender argument
    values in request queue. It works together with PrerenderMiddleware downloader
    middleware.
    """
    local_values_key = '_prerender_local_values'

    def process_spider_output(self, response, result, spider):
        for el in result:
            if isinstance(el, scrapy.Request):
                yield self._process_request(el, spider)
            else:
                yield el

    def process_start_requests(self, start_requests, spider):
        if not hasattr(spider, 'state'):
            spider.state = {}
        spider.state.setdefault(self.local_values_key, {})  # fingerprint => value dict

        for req in start_requests:
            yield self._process_request(req, spider)

    def _process_request(self, request, spider):
        """
        Replace requested meta['prerender']['args'] values with their fingerprints.
        This allows to store values only once in request queue, which helps
        with disk queue size.

        Downloader middleware should restore the values from fingerprints.
        """
        if 'prerender' not in request.meta:
            return request

        if '_replaced_args' in request.meta['prerender']:
            # don't process re-scheduled requests
            # XXX: does it work as expected?
            warnings.warn("Unexpected request.meta['prerender']['_replaced_args']")
            return request

        request.meta['prerender']['_replaced_args'] = []
        cache_args = request.meta['prerender'].get('cache_args', [])
        args = request.meta['prerender'].setdefault('args', {})

        for name in cache_args:
            if name not in args:
                continue
            value = args[name]
            fp = 'LOCAL+' + json_based_hash(value)
            spider.state[self.local_values_key][fp] = value
            args[name] = fp
            request.meta['prerender']['_replaced_args'].append(name)

        return request


class PrerenderMiddleware(object):
    """
    Scrapy downloader and spider middleware that passes requests
    through Prerender when 'prerender' Request.meta key is set.

    This middleware also works together with PrerenderDeduplicateArgsMiddleware
    spider middleware to allow not to store duplicate Prerender argument values
    in request queue and not to send them multiple times to Prerender
    (the latter requires Prerender 2.1+).
    """
    default_prerender_url = 'http://127.0.0.1:8050'
    default_endpoint = "render.json"
    prerender_extra_timeout = 5.0
    default_policy = SlotPolicy.PER_DOMAIN
    rescheduling_priority_adjust = +100
    retry_498_priority_adjust = +50
    remote_keys_key = '_prerender_remote_keys'

    def __init__(self, crawler, prerender_base_url, slot_policy, log_400):
        self.crawler = crawler
        self.prerender_base_url = prerender_base_url
        self.slot_policy = slot_policy
        self.log_400 = log_400
        self.crawler.signals.connect(self.spider_opened, signals.spider_opened)

    @classmethod
    def from_crawler(cls, crawler):
        prerender_base_url = crawler.settings.get('PRERENDER_URL',
                                               cls.default_prerender_url)
        log_400 = crawler.settings.getbool('PRERENDER_LOG_400', True)
        slot_policy = crawler.settings.get('PRERENDER_SLOT_POLICY',
                                           cls.default_policy)
        if slot_policy not in SlotPolicy._known:
            raise NotConfigured("Incorrect slot policy: %r" % slot_policy)

        return cls(crawler, prerender_base_url, slot_policy, log_400)

    def spider_opened(self, spider):
        if not hasattr(spider, 'state'):
            spider.state = {}

        # local fingerprint => key returned by prerender
        spider.state.setdefault(self.remote_keys_key, {})

    @property
    def _argument_values(self):
        key = PrerenderDeduplicateArgsMiddleware.local_values_key
        return self.crawler.spider.state[key]

    @property
    def _remote_keys(self):
        return self.crawler.spider.state[self.remote_keys_key]

    def process_request(self, request, spider):
        if 'prerender' not in request.meta:
            return

        if request.method not in {'GET', 'POST'}:
            logger.warning(
                "Currently only GET and POST requests are supported by "
                "PrerenderMiddleware; %(request)s will be handled without Prerender",
                {'request': request},
                extra={'spider': spider}
            )
            return request

        if request.meta.get("_prerender_processed"):
            # don't process the same request more than once
            return

        prerender_options = request.meta['prerender']
        request.meta['_prerender_processed'] = True

        slot_policy = prerender_options.get('slot_policy', self.slot_policy)
        self._set_download_slot(request, request.meta, slot_policy)

        args = prerender_options.setdefault('args', {})

        if '_replaced_args' in prerender_options:
            # restore arguments before sending request to the downloader
            load_args = {}
            save_args = []
            local_arg_fingerprints = {}
            for name in prerender_options['_replaced_args']:
                fp = args[name]
                # Use remote Prerender argument cache: if Prerender key
                # for a value is known then don't send the value to Prerender;
                # if it is unknown then try to save the value on server using
                # ``save_args``.
                if fp in self._remote_keys:
                    load_args[name] = self._remote_keys[fp]
                    del args[name]
                else:
                    save_args.append(name)
                    args[name] = self._argument_values[fp]

                local_arg_fingerprints[name] = fp

            if load_args:
                args['load_args'] = load_args
            if save_args:
                args['save_args'] = save_args
            prerender_options['_local_arg_fingerprints'] = local_arg_fingerprints

            del prerender_options['_replaced_args']  # ??

        args.setdefault('url', request.url)
        if request.method == 'POST':
            args.setdefault('http_method', request.method)
            # XXX: non-UTF8 request bodies are not supported now
            args.setdefault('body', request.body.decode('utf8'))

        if not prerender_options.get('dont_send_headers'):
            headers = scrapy_headers_to_unicode_dict(request.headers)
            if headers:
                args.setdefault('headers', headers)

        body = json.dumps(args, ensure_ascii=False, sort_keys=True, indent=4)
        # print(body)

        if 'timeout' in args:
            # User requested a Prerender timeout explicitly.
            #
            # We can't catch a case when user requested `download_timeout`
            # explicitly because a default value for `download_timeout`
            # is set by DownloadTimeoutMiddleware.
            #
            # As user requested Prerender timeout explicitly, we shouldn't change
            # it. Another reason not to change the requested Prerender timeout is
            # because it may cause a validation error on the remote end.
            #
            # But we can change Scrapy `download_timeout`: increase
            # it when it's too small. Decreasing `download_timeout` is not
            # safe.

            timeout_requested = float(args['timeout'])
            timeout_expected = timeout_requested + self.prerender_extra_timeout

            # no timeout means infinite timeout
            timeout_current = request.meta.get('download_timeout', 1e6)

            if timeout_expected > timeout_current:
                request.meta['download_timeout'] = timeout_expected

        endpoint = prerender_options.setdefault('endpoint', self.default_endpoint)
        prerender_base_url = prerender_options.get('prerender_url', self.prerender_base_url)
        prerender_url = urljoin(prerender_base_url, endpoint)

        headers = Headers({'Content-Type': 'application/json'})
        headers.update(prerender_options.get('prerender_headers', {}))
        new_request = request.replace(
            url=prerender_url,
            method='POST',
            body=body,
            headers=headers,
            priority=request.priority + self.rescheduling_priority_adjust
        )
        self.crawler.stats.inc_value('prerender/%s/request_count' % endpoint)
        return new_request

    def process_response(self, request, response, spider):
        if not request.meta.get("_prerender_processed"):
            return response

        prerender_options = request.meta['prerender']
        if not prerender_options:
            return response

        # update stats
        endpoint = prerender_options['endpoint']
        self.crawler.stats.inc_value(
            'prerender/%s/response_count/%s' % (endpoint, response.status)
        )

        # handle save_args/load_args
        self._process_x_prerender_saved_arguments(request, response)
        if get_prerender_status(response) == 498:
            logger.debug("Got HTTP 498 response for {}; "
                         "sending arguments again.".format(request),
                         extra={'spider': spider})
            return self._498_retry_request(request, response)

        if prerender_options.get('dont_process_response', False):
            return response

        response = self._change_response_class(request, response)

        if self.log_400 and get_prerender_status(response) == 400:
            self._log_400(request, response, spider)

        return response

    def _change_response_class(self, request, response):
        from scrapy_prerender import PrerenderResponse, PrerenderTextResponse
        if not isinstance(response, (PrerenderResponse, PrerenderTextResponse)):
            # create a custom Response subclass based on response Content-Type
            # XXX: usually request is assigned to response only when all
            # downloader middlewares are executed. Here it is set earlier.
            # Does it have any negative consequences?
            respcls = responsetypes.from_args(headers=response.headers)
            if isinstance(response, TextResponse) and respcls is PrerenderResponse:
                # Even if the headers say it's binary, it has already
                # been detected as a text response by scrapy (for example
                # because it was decoded successfully), so we should not
                # convert it to PrerenderResponse.
                respcls = PrerenderTextResponse
            response = response.replace(cls=respcls, request=request)
        return response

    def _log_400(self, request, response, spider):
        from scrapy_prerender import PrerenderJsonResponse
        if isinstance(response, PrerenderJsonResponse):
            logger.warning(
                "Bad request to Prerender: %s" % response.data,
                {'request': request},
                extra={'spider': spider}
            )

    def _process_x_prerender_saved_arguments(self, request, response):
        """ Keep track of arguments saved by Prerender. """
        saved_args = get_prerender_headers(response).get(b'X-Prerender-Saved-Arguments')
        if not saved_args:
            return
        saved_args = parse_x_prerender_saved_arguments_header(saved_args)
        arg_fingerprints = request.meta['prerender']['_local_arg_fingerprints']
        for name, key in saved_args.items():
            fp = arg_fingerprints[name]
            self._remote_keys[fp] = key

    def _498_retry_request(self, request, response):
        """
        Return a retry request for HTTP 498 responses. HTTP 498 means
        load_args are not present on server; client should retry the request
        with full argument values instead of their hashes.
        """
        meta = copy.deepcopy(request.meta)
        local_arg_fingerprints = meta['prerender']['_local_arg_fingerprints']
        args = meta['prerender']['args']
        args.pop('load_args', None)
        args['save_args'] = list(local_arg_fingerprints.keys())

        for name, fp in local_arg_fingerprints.items():
            args[name] = self._argument_values[fp]
            # print('remote_keys before:', self._remote_keys)
            self._remote_keys.pop(fp, None)
            # print('remote_keys after:', self._remote_keys)

        body = json.dumps(args, ensure_ascii=False, sort_keys=True, indent=4)
        # print(body)
        request = request.replace(
            meta=meta,
            body=body,
            priority=request.priority+self.retry_498_priority_adjust
        )
        return request

    def _set_download_slot(self, request, meta, slot_policy):
        if slot_policy == SlotPolicy.PER_DOMAIN:
            # Use the same download slot to (sort of) respect download
            # delays and concurrency options.
            meta['download_slot'] = self._get_slot_key(request)

        elif slot_policy == SlotPolicy.SINGLE_SLOT:
            # Use a single slot for all Prerender requests
            meta['download_slot'] = '__prerender__'

        elif slot_policy == SlotPolicy.SCRAPY_DEFAULT:
            # Use standard Scrapy concurrency setup
            pass

    def _get_slot_key(self, request_or_response):
        return self.crawler.engine.downloader._get_slot_key(
            request_or_response, None
        )
